src/symbol/placement.js

```diff
@@ -83,6 +82,45 @@ export class RetainedQueryData {
     }
 }
 
+class CollisionGroups {
+    collisionGroups: { [layerId: ?string]: { ID: number, predicate?: any }};
+    maxGroupID: number;
+
+    constructor() {
+        this.maxGroupID = 0;
+        this.collisionGroups = {
+            undefined: {
+                ID: 0,
+                predicate: null
+            }
+        };
+    }
+
+    get(groupName?: string) {
+        if (groupName && this.maxGroupID === 0) {
+            // Keep the predicate null until the first collision
+            // group gets added to avoid overhead in the case
+            // everything's in the default group.
```

Comment:
Will this ever happen in the current implementation?  It looks to me like if `crossSourceCollisions` is set `false`, then every layer gets a named group, and otherwise, every layer uses the default group.

Reply:
Yeah, this is probably weird dead code to leave around if we're going to use the `crossSourceCollisions: bool` approach. We might try to add something to skip the expense if `crossSourceCollisions` is `true` and there's only one source, but it's probably not an important optimization.

Reply:
I changed `CollisionGroups` to explicitly work with source IDs now, which avoids the dead code and any worries about how to manage a "default" group.